File: WhatisExplorer_Lite/WhatisExplorerLite/Models/POI.swift
//
//  POI.swift
//  Whatis Explorer – Lite
//
//  POI Model - Compatible with backend structure
//

import Foundation
import CoreLocation

struct POI: Identifiable, Codable, Equatable {
    let id: String
    let name: String
    let description: String
    let lat: Double
    let lng: Double
    let zone: String
    let category: String
    let imageUrl: String?
    let coordStatus: String?
    
    // Computed properties
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }
    
    var location: CLLocation {
        CLLocation(latitude: lat, longitude: lng)
    }
    
    // Category icon mapping
    var iconName: String {
        switch category.lowercased() {
        case "monument": return "building.columns"
        case "church": return "building.columns.fill"
        case "marina": return "sailboat"
        case "beach": return "beach.umbrella"
        case "biological": return "leaf"
        case "wreck": return "ship"
        case "viewpoint": return "binoculars"
        case "village": return "house"
        case "event": return "calendar"
        case "restaurant": return "fork.knife"
        case "hotel": return "bed.double"
        case "museum": return "building.2"
        case "park": return "tree"
        case "harbor": return "anchor"
        case "lighthouse": return "light.beacon"
        case "cave": return "mountain.2"
        case "mountain": return "mountain.2.fill"
        case "lake": return "water.waves"
        case "river": return "water.waves.slash"
        case "villa": return "building"
        default: return "mappin.circle"
        }
    }
    
    var categoryDisplayName: String {
        switch category.lowercased() {
        case "monument": return "Monumento"
        case "church": return "Chiesa"
        case "marina": return "Marina"
        case "beach": return "Spiaggia"
        case "biological": return "Biologico"
        case "wreck": return "Relitto"
        case "viewpoint": return "Belvedere"
        case "village": return "Villaggio"
        case "event": return "Evento"
        case "restaurant": return "Ristorante"
        case "hotel": return "Hotel"
        case "museum": return "Museo"
        case "park": return "Parco"
        case "harbor": return "Porto"
        case "lighthouse": return "Faro"
        case "cave": return "Grotta"
        case "mountain": return "Montagna"
        case "lake": return "Lago"
        case "river": return "Fiume"
        case "villa": return "Villa"
        default: return "Altro"
        }
    }
    
    // Calculate distance from current location
    func distance(from location: CLLocation) -> CLLocationDistance {
        self.location.distance(from: location)
    }
    
    func formattedDistance(from location: CLLocation) -> String {
        let distance = self.distance(from: location)
        if distance < 1000 {
            return String(format: "%.0f m", distance)
        } else {
            return String(format: "%.1f km", distance / 1000)
        }
    }
    
    // Coordinate status
    var isCoordinateConfirmed: Bool {
        coordStatus == "confirmed"
    }
    
    var coordinateStatusBadge: String {
        switch coordStatus {
        case "confirmed": return "✓ Confermate"
        case "unconfirmed": return "? Da verificare"
        case "missing": return "✗ Mancanti"
        default: return "? Da verificare"
        }
    }
}

// MARK: - Backend Response Models
struct POIResponse: Codable {
    let _id: String
    let name: String
    let description: String?
    let lat: Double
    let lng: Double
    let zone: String
    let category: String?
    let imageUrl: String?
    let coordStatus: String?
    
    func toPOI() -> POI {
        POI(
            id: _id,
            name: name,
            description: description ?? "",
            lat: lat,
            lng: lng,
            zone: zone,
            category: category ?? "other",
            imageUrl: imageUrl,
            coordStatus: coordStatus
        )
    }
}

File: android/app/src/main/java/com/andaly/whatisexplorer/models/POI.kt
package com.andaly.whatisexplorer.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.Serializable

/**
 * POI Model - Compatible with backend structure
 * Based on iOS POI.swift
 */
@Parcelize
@Serializable
data class POI(
    val id: String,
    val name: String,
    val description: String,
    val lat: Double,
    val lng: Double,
    val zone: String,
    val category: String,
    val imageUrl: String? = null,
    val coordStatus: String? = null
) : Parcelable {
    
    // Category icon mapping (Material Icons)
    fun getIconName(): String {
        return when (category.lowercase()) {
            "monument" -> "apartment"
            "church" -> "church"
            "marina" -> "sailing"
            "beach" -> "beach_access"
            "biological" -> "eco"
            "wreck" -> "anchor"
            "viewpoint" -> "visibility"
            "village" -> "location_city"
            "event" -> "event"
            "restaurant" -> "restaurant"
            "hotel" -> "hotel"
            "museum" -> "museum"
            "park" -> "park"
            "harbor" -> "local_shipping"
            "lighthouse" -> "lighthouse"
            "cave" -> "landscape"
            "mountain" -> "landscape"
            "lake" -> "water"
            "river" -> "water"
            "villa" -> "villa"
            else -> "place"
        }
    }
    
    fun getCategoryDisplayName(): String {
        return when (category.lowercase()) {
            "monument" -> "Monumento"
            "church" -> "Chiesa"
            "marina" -> "Marina"
            "beach" -> "Spiaggia"
            "biological" -> "Biologico"
            "wreck" -> "Relitto"
            "viewpoint" -> "Belvedere"
            "village" -> "Villaggio"
            "event" -> "Evento"
            "restaurant" -> "Ristorante"
            "hotel" -> "Hotel"
            "museum" -> "Museo"
            "park" -> "Parco"
            "harbor" -> "Porto"
            "lighthouse" -> "Faro"
            "cave" -> "Grotta"
            "mountain" -> "Montagna"
            "lake" -> "Lago"
            "river" -> "Fiume"
            "villa" -> "Villa"
            else -> "Altro"
        }
    }
    
    fun isCoordinateConfirmed(): Boolean {
        return coordStatus == "confirmed"
    }
    
    fun getCoordinateStatusBadge(): String {
        return when (coordStatus) {
            "confirmed" -> "✓ Confermate"
            "unconfirmed" -> "? Da verificare"
            "missing" -> "✗ Mancanti"
            else -> "? Da verificare"
        }
    }
}

/**
 * Backend Response Models - Based on iOS MobilePOIsResponse
 */
@Serializable
data class MobilePOIsResponse(
    val success: Boolean,
    val zone: MobileZoneInfo,
    val pois: List<MobilePOIResponse>,
    val totalPOIs: Int
)

@Serializable
data class MobileZoneInfo(
    val id: String,
    val name: String,
    val description: String? = null,
    val geographicArea: MobileGeographicArea? = null
)

@Serializable
data class MobileGeographicArea(
    val id: String,
    val name: String,
    val displayName: String
)

@Serializable
data class MobilePOIResponse(
    val id: String,
    val name: String,
    val description: String? = null,
    val lat: Double,
    val lng: Double,
    val category: String? = null,
    val semanticCategory: String? = null,
    val imageUrl: String? = null,
    val customIcon: String? = null,
    val source: String? = null,
    val extraInfo: MobilePOIExtraInfo
) {
    fun toPOI(zoneId: String): POI {
        return POI(
            id = id,
            name = name,
            description = description ?: "",
            lat = lat,
            lng = lng,
            zone = zoneId,
            category = category ?: semanticCategory ?: "other",
            imageUrl = imageUrl,
            coordStatus = null
        )
    }
}

@Serializable
data class MobilePOIExtraInfo(
    val rating: Double = 0.0,
    val accessibility: String = "public",
    val tags: List<String> = emptyList()
)

